src/main/java/org/cryptacular/CiphertextHeaderV2.java

```diff
@@ -0,0 +1,307 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.cryptacular;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import javax.crypto.SecretKey;
+import org.bouncycastle.crypto.digests.SHA256Digest;
+import org.bouncycastle.crypto.macs.HMac;
+import org.cryptacular.util.ByteUtil;
+
+/**
+ * Cleartext header prepended to ciphertext providing data required for decryption.
+ *
+ * <p>Data format:</p>
+ *
+ * <pre>
+     +---------+---------+---+----------+-------+------+
+     | Version | KeyName | 0 | NonceLen | Nonce | HMAC |
+     +---------+---------+---+----------+-------+------+
+     |                                                 |
+     +--- 4 ---+--- x ---+ 1 +--- 1 ----+-- y --+- 32 -+
```

Comment:
@dhawes curious what you think about using NULL as the delimiter here.

Reply:
Assuming KeyName contains no NULL characters, I don't see much issue with it in practice.

An alternative could be to add length fields to specify the number of bytes to be read, e.g. KeyNameLength.

Reply:
I believe this implementation was intended to improve on v1 which did use a keyNameLength.

Reply:
That is correct.